File: Data Structures & Algorithms/combination-target-sum-ii/submission-1.java
class Solution {
    public List<List<Integer>> combinationSum2(int[] nums, int target) {
        Arrays.sort(nums);
        List<List<Integer>> solutions = new ArrayList<>();
        backtrack(nums, target, 0, new ArrayList<>(), 0, solutions);
        return solutions;
    }

    public void backtrack(int[] nums, int target, int start, List<Integer> current, int sum, List<List<Integer>> ans) {
        if (sum == target) {
            ans.add(new ArrayList<>(current));
            return;
        }
        if (sum > target) return;
        for (int i = start; i < nums.length; i++) {
            if (i > start && nums[i] == nums[i - 1]) continue;
            current.add(nums[i]);
            sum += nums[i];
            backtrack(nums, target, i + 1, current, sum, ans);
            sum -= nums[i];
            current.remove(current.size() - 1);
        }
    }
}

File: Data Structures & Algorithms/subsets/submission-1.cpp
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> ans;
        vector<int> current;
        backtrack(ans, current, nums, 0);
        return ans;
    }

    void backtrack(vector<vector<int>>& ans, vector<int>& current, vector<int>& nums, int start) {
        ans.push_back(std::vector(current));
        for (int i = start; i < nums.size(); i++) {
            current.push_back(nums[i]);
            backtrack(ans, current, nums, i + 1);
            current.pop_back();
        }
    }
};
